### ISSUE TITLE
Byte-compilation error in Emacs 29

### ISSUE THREAD

**Issue by akirak** [alphapapa/org-bookmark-heading#18]
Emacs 29 now throws a byte-compile error:

> org-bookmark-heading.el:176:6: Error: Unknown map pattern: (map 'filename 'outline-path 'id 'front-context-string 'indirectp)

This is because now `pcase.el` loads the pattern definition from `pcase-macroexpander` property of each symbol:

```emacs-lisp
(defun pcase--get-macroexpander (s)
  "Return the macroexpander for pcase pattern head S, or nil."
  (get s 'pcase-macroexpander))
```

`pcase-macroexpander` property for `map` is defined in `map.el`. Thus, you can fix the error by loading `map.el` ~before you use `map` pattern~ at compile time:

```emacs-lisp
(eval-when-compile
  ;; Support map pattern in pcase
  (require 'map))
```

**Comment by alphapapa** [comment]
Hi Akira,

Thanks for reporting this and filing the PR to fix it.

Would you please consider reporting this as an Emacs bug as well?  This seems like a backward-incompatibility, or a regression of sorts, and it's likely that the Emacs maintainers aren't aware of how that change will affect downstream packages, i.e. if they suddenly stop byte-compiling correctly on Emacs 29, that seems like a problem that should be avoided if at all possible.

**Comment by akirak** [comment]
> Would you please consider reporting this as an Emacs bug as well? 

I am not sure if this should be considered a bug. `pcase-defmacro` itself has existed for 8 years and `map` pattern was defined using it since its introduction. Maybe the error was caused by the byte compiler which is now stricter than before. After all, `map` is one of the additional patterns according to the documentation, so it may make sense to load it explicitly. 

Nonetheless, it may worth discussing with the maintainers. Maybe I'll report it later.

**Comment by akirak** [comment]
I have reported the issue to bug-gnu-emacs. Thank you for the direction!

**Comment by alphapapa** [comment]
Thanks.  Would you please link the report here?

**Comment by akirak** [comment]
The discussion is about to end here:

> Maybe it started with
> | 85555ad3b79 ; Require map only during compilation
> | Philip Kaludercic <philipk@posteo.net> 2022-10-23

https://lists.gnu.org/archive/html/bug-gnu-emacs/2022-12/msg00757.html

Michael refers to the following commit, at which map is required only during compilation in `package-vc.el`:

https://github.com/emacs-mirror/emacs/commit/85555ad3b79806f662819c74810d9db26108ef9f

**Comment by akirak** [comment]
Hello, 

The Emacs maintainers have concluded that the issue is not a bug of Emacs. The issue has been closed, and you will have to merge #19. The closing message is attached below.

Thanks.

> Your bug report
> 
> #59900: 30.0.50; map pattern in pcase causes a byte-compile error
> 
> which was filed against the emacs package, has been closed.
> 
> The explanation is attached below, along with your original report.
> If you require more details, please reply to 59900@debbugs.gnu.org.
> 
> -- 
> 59900: https://debbugs.gnu.org/cgi/bugreport.cgi?bug=59900
> GNU Bug Tracking System
> Contact help-debbugs@gnu.org with problems
> 
> 
> 
> ---------- Forwarded message ----------
> From: Eli Zaretskii <eliz@gnu.org>
> To: Michael Heerdegen <michael_heerdegen@web.de>
> Cc: akira.komamura@gmail.com, 59900-done@debbugs.gnu.org
> Bcc: 
> Date: Sat, 10 Dec 2022 09:53:42 +0200
> Subject: Re: bug#59900: 30.0.50; map pattern in pcase causes a byte-compile error
> > From: Michael Heerdegen <michael_heerdegen@web.de>
> > Cc: akira.komamura@gmail.com,  59900@debbugs.gnu.org
> > Date: Sat, 10 Dec 2022 02:01:35 +0100
> >

**Comment by stefan2904** [comment]
The warning/error also appears in Emacs 28.1, by the way.

> Warning (comp): /home/stefan/.emacs.d/elpa/org-bookmark-heading-20221022.1422/org-bookmark-heading.el: Error: error Unknown map pattern: (map filename outline-path id front-context-string indirectp) Disable showing Disable logging
